ui/src/server_axum.rs

```diff
@@ -0,0 +1,511 @@
+        GenerateLabels, SuccessDetails,
+    },
+    sandbox::{self, Channel, Sandbox},
+    CachingSnafu, ClippyRequest, ClippyResponse, CompilationSnafu, CompileRequest, CompileResponse,
+    Config, Error, ErrorJson, EvaluateRequest, EvaluateResponse, EvaluationSnafu, ExecuteRequest,
+    ExecuteResponse, ExecutionSnafu, ExpansionSnafu, FormatRequest, FormatResponse,
+    FormattingSnafu, GhToken, GistCreationSnafu, GistLoadingSnafu, InterpretingSnafu, LintingSnafu,
+    MacroExpansionRequest, MacroExpansionResponse, MetaCratesResponse, MetaGistCreateRequest,
+    MetaGistResponse, MetaVersionResponse, MetricsToken, MiriRequest, MiriResponse, Result,
+    SandboxCreationSnafu, SpawnBlockingSandboxSnafu, ONE_HOUR, SANDBOX_CACHE_TIME_TO_LIVE,
+};
+use async_trait::async_trait;
+use axum::{
+    extract::{self, Extension, Path, TypedHeader},
+    handler::Handler,
+    headers::{authorization::Bearer, Authorization},
+    http::{header, uri::PathAndQuery, HeaderValue, Method, Request, StatusCode, Uri},
+    response::IntoResponse,
+    routing::{get, get_service, post, MethodRouter},
+    AddExtensionLayer, Json, Router,
+};
+use axum_extra::middleware;
+use snafu::{IntoError, ResultExt};
+use std::{
+    convert::{TryFrom, TryInto},
+    mem, path,
+    sync::Arc,
+    time::Instant,
+};
+use tokio::sync::Mutex;
+use tower_http::{
+    cors::{self, CorsLayer},
+    services::ServeDir,
+    set_header::SetResponseHeader,
+    trace::TraceLayer,
+};
+
+const MAX_AGE_ONE_DAY: HeaderValue = HeaderValue::from_static("public, max-age=86400");
+const MAX_AGE_ONE_YEAR: HeaderValue = HeaderValue::from_static("public, max-age=31536000");
+
+#[tokio::main]
+pub(crate) async fn serve(config: Config) {
+    let root_files = static_file_service(config.root_path(), MAX_AGE_ONE_DAY);
+    let asset_files = static_file_service(config.asset_path(), MAX_AGE_ONE_YEAR);
+    let rewrite_help_as_index = middleware::from_fn(rewrite_help_as_index);
+
+    let mut app = Router::new()
+        .fallback(root_files)
+        .nest("/assets", asset_files)
+        .layer(rewrite_help_as_index)
+        .route("/evaluate.json", post(evaluate))
+        .route("/compile", post(compile))
+        .route("/execute", post(execute))
+        .route("/format", post(format))
+        .route("/clippy", post(clippy))
+        .route("/miri", post(miri))
+        .route("/macro-expansion", post(macro_expansion))
+        .route("/meta/crates", get_or_post(meta_crates))
+        .route("/meta/version/stable", get_or_post(meta_version_stable))
+        .route("/meta/version/beta", get_or_post(meta_version_beta))
+        .route("/meta/version/nightly", get_or_post(meta_version_nightly))
+        .route("/meta/version/rustfmt", get_or_post(meta_version_rustfmt))
+        .route("/meta/version/clippy", get_or_post(meta_version_clippy))
+        .route("/meta/version/miri", get_or_post(meta_version_miri))
+        .route("/meta/gist/", post(meta_gist_create))
```

Comment:
Is the trailing slash intended here? None of the other routes have it.

If someone requests `/meta/gist` axum will send a redirect to `/meta/gist/`.

Reply:
Ah, that can probably be changed back now. That was how I found https://github.com/tokio-rs/axum/issues/681